src/snowflake/cli/api/utils/definition_rendering.py

```diff
@@ -0,0 +1,196 @@
+        """
+        current_dict_level = context
+        for i, var in enumerate(self._vars_chain):
+            if i == len(self._vars_chain) - 1:
+                current_dict_level[var] = value
+            else:
+                current_dict_level.setdefault(var, {})
+                current_dict_level = current_dict_level[var]
+
+    def read_from_context(self, context):
+        """
+        Takes a context dict as input.
+
+        Traverse through the multi-level dictionary to the location where this variable goes to.
+        Returns the value in that location.
+
+        Raise UndefinedError if the variable is None or not found.
+        """
+        current_dict_level = context
+        for key in self._vars_chain:
+            if (
+                not isinstance(current_dict_level, dict)
+                or key not in current_dict_level
+            ):
+                raise UndefinedError(
+                    f"Could not find template variable {self.get_key()}"
+                )
+            current_dict_level = current_dict_level[key]
+
+        value = current_dict_level
+        if value == None or isinstance(value, dict) or isinstance(value, list):
+            raise UndefinedError(
+                f"Template variable {self.get_key()} does not contain a valid value"
+            )
+
+        return value
+
+    def __hash__(self):
+        return hash(self.get_key())
+
+    def __eq__(self, other):
+        return self.get_key() == other.get_key()
+
+
+def _get_referenced_vars(ast_node, current_attr_chain: list[str] = []) -> set[Variable]:
+    """
+    Traverse Jinja AST to find the variable chain referenced by the template.
+    A variable like ctx.env.test is internally represented in the AST tree as
+    Getattr Node (attr='test') -> Getattr Node (attr='env') -> Name Node (name='ctx')
+    """
+    all_referenced_vars = set()
+    if isinstance(ast_node, nodes.Getattr):
+        current_attr_chain = [getattr(ast_node, "attr")] + current_attr_chain
+    elif isinstance(ast_node, nodes.Name):
+        current_attr_chain = [getattr(ast_node, "name")] + current_attr_chain
+        all_referenced_vars.add(Variable(current_attr_chain))
+
+    for child_node in ast_node.iter_child_nodes():
+        all_referenced_vars.update(_get_referenced_vars(child_node, current_attr_chain))
+
+    return all_referenced_vars
+
+
+def _get_referenced_vars_from_str(env: Environment, template_str: str) -> set[Variable]:
+    ast = env.parse(template_str)
+    return _get_referenced_vars(ast)
+
+
+def _build_dependency_graph(
+    env: Environment, all_vars: set[Variable], context
+) -> Graph[Variable]:
+    dependencies_graph = Graph[Variable]()
+    for variable in all_vars:
+        dependencies_graph.add(Node[Variable](key=variable.get_key(), data=variable))
+
+    for variable in all_vars:
+        if variable.is_env_var() and variable.get_env_var_name() in os.environ:
```

Comment:
Nit: I feel like any sort of env dict should be able to be passed in, rather than having to use `os.environ`. Maybe a future refactor when we need it, but it could also simplify testing.

Reply:
I will keep it as is for now - I have the another task that might involve this kind of refactor, so I will try to do that in there.